web-ui/router.ml

```diff
@@ -8,199 +8,194 @@ let loader root path _request =
+    Dream.get "/gitlab/:org" (fun request ->
+        Controller.Gitlab.list_repos ~org:(Dream.param request "org") gitlab);
+    Dream.get "/gitlab/:org/:repo" (fun request ->
+        Controller.Gitlab.list_refs
+          ~org:(Dream.param request "org")
+          ~repo:(Dream.param request "repo")
+          gitlab);
+    Dream.get "/gitlab/:org/:repo/commit/:hash" (fun request ->
+        Controller.Gitlab.list_steps
+          ~org:(Dream.param request "org")
+          ~repo:(Dream.param request "repo")
+          ~hash:(Dream.param request "hash")
+          request gitlab);
+    Dream.get "/gitlab/:org/:repo/commit/:hash/variant/:variant" (fun request ->
+        Controller.Gitlab.show_step
+          ~org:(Dream.param request "org")
+          ~repo:(Dream.param request "repo")
+          ~hash:(Dream.param request "hash")
+          ~variant:(Dream.param request "variant")
+          request gitlab);
+    Dream.post "/gitlab/:org/:repo/commit/:hash/variant/:variant/rebuild"
+      (fun request ->
+        Dream.form request >>= function
+        | `Ok _ ->
+            Controller.Gitlab.rebuild_step
+              ~org:(Dream.param request "org")
+              ~repo:(Dream.param request "repo")
+              ~hash:(Dream.param request "hash")
+              ~variant:(Dream.param request "variant")
+              request gitlab
+        | _ ->
+            Dream.log "Form validation failed";
+            Dream.empty `Bad_Request);
+    Dream.post "/gitlab/:org/:repo/commit/:hash/cancel" (fun request ->
+        Dream.form request >>= function
+        | `Ok _ ->
+            Controller.Gitlab.cancel_steps
+              ~org:(Dream.param request "org")
+              ~repo:(Dream.param request "repo")
+              ~hash:(Dream.param request "hash")
+              request gitlab
+        | _ ->
+            Dream.log "Form validation failed";
+            Dream.empty `Bad_Request);
+    Dream.post "/gitlab/:org/:repo/commit/:hash/rebuild-failed" (fun request ->
+        Dream.form request >>= function
+        | `Ok _ ->
+            Controller.Gitlab.rebuild_steps ~rebuild_failed_only:true
+              ~org:(Dream.param request "org")
+              ~repo:(Dream.param request "repo")
+              ~hash:(Dream.param request "hash")
+              request gitlab
+        | _ ->
+            Dream.log "Form validation failed";
+            Dream.empty `Bad_Request);
+    Dream.post "/gitlab/:org/:repo/commit/:hash/rebuild-all" (fun request ->
+        Dream.form request >>= function
+        | `Ok _ ->
+            Controller.Gitlab.rebuild_steps ~rebuild_failed_only:false
+              ~org:(Dream.param request "org")
+              ~repo:(Dream.param request "repo")
+              ~hash:(Dream.param request "hash")
+              request gitlab
+        | _ ->
+            Dream.log "Form validation failed";
+            Dream.empty `Bad_Request);
+  ]
+
 let create ~github ~gitlab =
+  let ( ++ ) = List.append in
```

Comment:
Why not use `@` instead?

Reply:
Haskellers in hiding!

Reply:
Caught! I really want an ad-hoc way to introduce infix operators.